### PR TITLE
Os 176 Bug while creating index for databaseProvider 

### PR DESCRIPTION
**Pull request by pritha-tarento**
Issue below:
1.Read API neo4jGraph throwing exception: Connection to the database was lost because someone called `interrupt()
2. Index Field property values were getting duplicating while adding new index field value.

As a part of this PR:
Introduced Indexer which  included 
+ createIndex(), setIndexFields(), setUniqueIndexFields
  updateIndices(), - createNonUniqueIndex(), - createUniqueIndex()

Introduced IndexHelper which included 2 methods
+ isIndexPresent() , getNewFields()

Added 2 methods in EntityParenter:
- asyncAddIndex() 
ensureIndexExists() : checks isIndexPresent()  and calls for asyncAddIndex()

### CODE REVIEW

File: java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java [hunk 1]
```diff
@@ -238,18 +238,35 @@ private void ensureIndexExists(DatabaseProvider dbProvider, Graph graph, String
      * @return
      */
     private boolean indexFieldsExists(Vertex parentVertex, List<String> fields) {
+        boolean contains = false;
+        if (parentVertex.property(Constants.INDEX_FIELDS).isPresent()) {
+            for (String field : fields) {
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java]: Can you get rid of for loop and simply use `contains = fields.contains(parentVertex)` Overall I see not much difference between indexFieldsExists and the fieldExists - both achieve the same thing and there is a lot of code here. Please reduce.

File: java/registry/src/main/java/io/opensaber/registry/sink/Neo4jGraphProvider.java [hunk 2]
```diff
@@ -26,6 +26,8 @@
     private boolean profilerEnabled;
     private DBConnectionInfo connectionInfo;
     private Neo4jIdProvider neo4jIdProvider = new Neo4jIdProvider();
+    private Neo4JGraph neo4jGraph;
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/sink/Neo4jGraphProvider.java]: This is disallowed. Neo4JGraph we know cannot be persisted for more than 1 api request. It must be within methods only.

File: java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java (hunk 1 above)
Comment (pritha-tarento) [reply java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java]: indexFieldsExists (for list) uses  fieldExists(for a single field), there was one line repetation I removed.
String:contains() method is not efficient so I have used -Arrays.stream(indexFields).anyMatch(field::equals)

Comment (pritha-tarento) [reply java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java]: Done.

File: java/registry/src/main/java/io/opensaber/registry/sink/Neo4jGraphProvider.java (hunk 2 above)
Comment (pritha-tarento) [reply java/registry/src/main/java/io/opensaber/registry/sink/Neo4jGraphProvider.java]: Done

File: java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java [hunk 3]
```diff
@@ -201,28 +202,26 @@ public String addEntity(String jsonString) throws Exception {
...
             if (!indexFieldsExists(parentVertex, indexFields)){
-                dbProvider.createIndex(label, indexFields);
                 setPropertyValuesOnParentVertex(parentVertex, indexFields);
+                dbProvider.createIndex(label, indexFields);
 
             }
             if(!indexFieldsExists(parentVertex, indexUniqueFields)){
-                dbProvider.createUniqueIndex(label, indexUniqueFields);
                 setPropertyValuesOnParentVertex(parentVertex, indexUniqueFields);
+                dbProvider.createUniqueIndex(label, indexUniqueFields);
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java]: Why is this moved? What happens if we set the property, but failed to create index for whatever reason?

File: java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java [hunk 4]
```diff
@@ -239,13 +238,12 @@ private void ensureIndexExists(DatabaseProvider dbProvider, Graph graph, String
...
     private boolean indexFieldsExists(Vertex parentVertex, List<String> fields) {
         boolean contains = false;
-        if (parentVertex.property(Constants.INDEX_FIELDS).isPresent()) {
-            for (String field : fields) {
-                contains = fieldExists(parentVertex, field);
-                if(!contains)
-                    break;
-            }
+        for (String field : fields) {
+            contains = fieldExists(parentVertex, field);
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java]: isn't this confusing? There is a list of fields and even if one of them doesn't exist, we try recreate index on all the list of fields. Why can't it be just the one that doesn't exist?

File: java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java (hunk 3 above)
Comment (pritha-tarento) [reply java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java]: setProperty commit will happens with createIndex trx hence if createIndex fails the commit for setProperty also rolls backs.

File: java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java (hunk 4 above)
Comment (pritha-tarento) [reply java/registry/src/main/java/io/opensaber/registry/service/impl/RegistryServiceImpl.java]: Yes we can, Done.

File: java/registry/src/main/java/io/opensaber/registry/util/IndexHelper.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                return isIndexPresent;
+
+            List<String> indexUniqueFields = definition.getOsSchemaConfiguration().getUniqueIndexFields();
+            int uniqueIndexSize = (indexUniqueFields.size() == 0) ? 0
+                    : getNewFields(parentVertex, indexUniqueFields, true).size();
+            if (uniqueIndexSize > 0)
+                return isIndexPresent;
+
+        }
+        definitionIndexMap.put(defTitle, true);
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/util/IndexHelper.java]: We must turn this true from outside...like in the next line of ensureIndexExists().

File: java/registry/src/main/java/io/opensaber/registry/util/IndexHelper.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+     * 
+     * @param parentVertex
+     * @param definition
+     * @return
+     */
+    public boolean isIndexPresent(Vertex parentVertex, Definition definition) {
+        boolean isIndexPresent = false;
+        String defTitle = definition.getTitle();
+        logger.info("isIndexPresent flag for {}: {}", defTitle, definitionIndexMap.get(defTitle));
+        if (!definitionIndexMap.get(defTitle)) {
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/util/IndexHelper.java]: Please check why can't this be simplified to just this -`return definitionIndexMap.getOrDefault(defTitle, false);`

File: java/registry/src/main/java/io/opensaber/registry/util/Indexer.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+     * property
+     * 
+     * @param parentVertex
+     * @param values
+     * @param isUnique
+     */
+    private void updateIndices(Vertex parentVertex, List<String> values, boolean isUnique) {
+        String propertyName = isUnique ? Constants.UNIQUE_INDEX_FIELDS : Constants.INDEX_FIELDS;
+
+        if (values.size() > 0) {
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/util/Indexer.java]: We must have this as a function in VertexWriter ideally...but its ok. I'm just leaving my thought here.

File: java/registry/src/main/java/io/opensaber/registry/util/EntityParenter.java [hunk 8]
```diff
@@ -131,4 +140,60 @@ public Vertex getKnownParentVertex(String definition, String shardId) {
...
+     * 
+     * @param dbProvider
+     * @param parentVertex
+     * @param definition
+     */
+    public void ensureIndexExists(DatabaseProvider dbProvider, Vertex parentVertex, Definition definition) {
+        try{
+            if(!indexHelper.isIndexPresent(parentVertex, definition)){
+                asyncAddIndex(dbProvider, parentVertex, definition);
+            }
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/util/EntityParenter.java]: After adding the index here, set the defintion->bool map to true.
